pep-0727.rst

```diff
@@ -0,0 +1,516 @@
+PEP: 727
+Title: Documentation Metadata in Typing
+Author: Sebastián Ramírez <tiangolo@gmail.com>
+Sponsor: Jelle Zijlstra <jelle.zijlstra@gmail.com>
+Discussions-To:
+Status: Draft
+Type: Standards Track
+Topic: Typing
+Content-Type: text/x-rst
+Created: 28-Aug-2023
+Python-Version: 3.13
+Post-History:
+
+
+Abstract
+========
+
+This document proposes a way to complement docstrings to add additional documentation
+to Python symbols using type annotations with ``Annotated`` (in class attributes,
+function and method parameters, return values, and variables).
+
+
+Motivation
+==========
+
+The current standard method of documenting code APIs in Python is using docstrings.
+But there's no standard way to document parameters in docstrings.
+
+There are several pseudo-standards for the format in these docstrings, and new
+pseudo-standards can appear easily: numpy, Google, Keras, reST, etc.
+
+All these formats are some specific syntax inside a string. Because of this, when
+editing those docstrings, editors can't easily provide support for autocompletion,
+inline errors for broken syntax, etc.
+
+Editors don't have a way to support all the possible micro languages in the docstrings
+and show nice user interfaces when developers use libraries with those different
+formats. They could even add support for some of the syntaxes, but probably not all,
+or not completely.
+
+Because the docstring is in a different place in the code than the actual parameters
+and it requires duplication of information (the parameter name) the information about
+a parameter is easily in a place in the code quite far away from the declaration of
+the actual parameter. This means it's easy to refactor a function, remove a parameter,
+and forget to remove its docs. The same happens when adding a new parameter: it's easy
+to forget to add the docstring for it.
+
+Editors can't check or ensure the consistency of the parameters and their docs.
+
+It's not possible to programatically test these docstrings, for example to ensure
+documentation consistency for the parameters across several similar functions, at
+least not without implementing a custom syntax parser.
```

Comment:
Commenting on the replacement section (#3315):


> And because of this same duplication of information (the parameter name) editors and
other tools would need to have complex custom logic to check or ensure the
consistency of the parameters in the signature and in their docstring, or they
would simply not be able to support that.
>
> Additionally, it would be difficult to robustly parse varying existing docstring
conventions to programatically get the documentation for each individual parameter
or variable at runtime. This would be useful, for example,
for testing the contents of each parameter's documentation, to ensure consistency
across several similar functions, or to extract and expose that same parameter
documentation in some other way (e.g. an API, a CLI, etc).

This should be in the present tense, not the future conditional. Some editors and linters do support docstring inspection currently -- what you're really wanting is to expose this data to the runtime in a structured format.